cmd/update-main.go

```diff
@@ -235,20 +256,33 @@ func getDownloadURL() (downloadURL string) {
 	return minioReleaseURL + "minio"
 }
 
-func getUpdateInfo(timeout time.Duration, mode string) (older time.Duration, downloadURL string, err error) {
-	currentReleaseTime, err := GetCurrentReleaseTime()
+// getUpdateInfo - compares latest available official release with the
+// version of current server process; returns staleness of current
+// release and download URL of latest release.
+func getUpdateInfo(mode string, timeout time.Duration) (
+	older time.Duration, downloadURL string, err error) {
+
+	u := updateCheckParams{
+		version:            Version,
+		serverBinaryPath:   os.Args[0],
+		cgroupFile:         "/proc/self/cgroup",
+		releaseChecksumURL: minioReleaseURL + "minio.shasum",
+		serverMode:         mode,
+	}
```

Comment:
Yes, other than `mode` all fields are constants. If you see the previous code, this is clear there too (so why did the earlier code become complex anyway?!) - the reason for putting constant-like values as argument(s) to various functions is to be able to write tests. If these values are constants or are hard-coded then it is not possible to pass different values for the purposes of tests. See for example the test for `isDockerEnv()` - it is quite difficult to write it if the file path string "constant" `/proc/self/cgroup` is hard-coded - but clearly the test is useful and important here and explains why it is implemented in the way it is.

The problem with global variables (even those that do not change even during the lifetime of the process) is also related to writing tests. Global values impact multiple tests at the same time and so the test author needs to clearly understand which global variables are used or are affected by a function/code that is to be tested - the test author needs to ensure that before the test they are appropriately set and after the test cleaned up so it doesnt affect subsequent tests. This is generally a pain. This was discussed quite a lot in BLR office some months back with @balamurugana and @krisis - the problem with globals and why they need to be eliminated.

IMO, this is a clean implementation - serving the purpose of maintainability and testability.